File: lee/src/main/java/org/example/lee/题目/二叉树/b_从前序与中序遍历序列构造二叉树.go
package 二叉树

var num int

func buildTree(preorder []int, inorder []int) *TreeNode {
	//对于中序遍历 实际上有一个分治的思想
	//不断从root结点进行拆分
	//然后前序遍历从左到右，可以保证每次获得的都是左子树
	m := make(map[int]int)
	num = 0
	for k, v := range inorder {
		m[v] = k
	}
	return buildTreeTool(preorder, inorder, 0, len(inorder)-1, m)

}

func buildTreeTool(preorder []int, inorder []int, x, y int, m map[int]int) *TreeNode {
	if x > y {
		return nil
	}
	root := &TreeNode{
		Val:   preorder[num],
		Left:  nil,
		Right: nil,
	}
	mid := m[preorder[num]]
	num++
	root.Left = buildTreeTool(preorder, inorder, x, mid-1, m)
	root.Right = buildTreeTool(preorder, inorder, mid+1, y, m)
	return root
}

File: lee/src/main/java/org/example/lee/题目/二叉树/b_从前序与中序遍历序列构造二叉树.java
package org.example.lee.题目.二叉树;

import java.util.HashMap;

import org.example.lee.model.TreeNode;

/**
 * @author 刘家辉
 * @date 2024/01/11
 */
public class b_从前序与中序遍历序列构造二叉树 {

	public int index = 0;

	public TreeNode buildTree(int[] preorder, int[] inorder) {
		HashMap<Integer, Integer> map = new HashMap<>();
		for (int i = 0; i < inorder.length; i++) {
			map.put(inorder[i], i);
		}
		return tool(preorder, 0, preorder.length - 1, map);
	}

	public TreeNode tool(int[] preorder, int left, int right, HashMap<Integer, Integer> map) {

		if (left > right || index >= preorder.length) {
			return null;
		}
		TreeNode head = new TreeNode();
		head.val = preorder[index++];
		if (left == right) {
			return head;
		}
		int i = map.get(head.val);
		head.left = tool(preorder, left, i - 1, map);
		head.right = tool(preorder, i + 1, right, map);
		return head;
	}


}
